### ISSUE TITLE
Please add TaiChi Network

### ISSUE THREAD

**Issue by zhangzhongnan928** [AlphaWallet/alpha-wallet-ios#2489]
https://github.com/Taichi-Network/docs/blob/master/sendPriveteTx_tutorial.md

**Comment by hboon** [comment]
What would this do?

Other than reducing frontrunning, does it make the transaction cheaper and faster to process?

Do we connect to this endpoint instead of Infura for Ethereum mainnet (at least for posting transactions, will have to check if it supports other functionality) ?

**Comment by zhangzhongnan928** [comment]
> @zhangzhongnan928 What would this do?
hide the TX before it is confirmed 

> Other than reducing frontrunning, does it make the transaction cheaper and faster to process?
no, not cheaper. might be faster, but not important.

> Do we connect to this endpoint instead of Infura for Ethereum mainnet (at least for posting transactions, will have to check if it supports other functionality) ?

No, just add it as a new network(although it is using the same chain ID as the mainnet)
cc @JamesSmartCell same for Andriod

**Comment by hboon** [comment]
@vladyslav-iosdev you want to give this a go? Remember to check 2 things:

* if the RPC server works for functionality besides posting transaction. eg. estimating the gas limit
* that `RPCServer` can work with duplicate chainIds. There might be code that constructs `RPCServer` instances from the chainId or somehow do a mapping from chainId -> `RPCServer`. We have to check that those still work correctly.

**Comment by zhangzhongnan928** [comment]
Please use the RPC address in my private message.

**Comment by vladyslav-iosdev** [comment]
> * that `RPCServer` can work with duplicate chainIds. There might be code that constructs `RPCServer` instances from the chainId or somehow do a mapping from chainId -> `RPCServer`. We have to check that those still work correctly.

@hboon  seems like there is a problem, how we could map right RPCServer if we only have only chainID? but we some additional check here.

**Comment by hboon** [comment]
How about we invent a different chainId (say `1999`, it's not listed in https://github.com/ethereum-lists/chains/tree/master/_data/chains) and assign that to TaiChi, so:

```
enum RPCServer... {
    var chainID: Int {
        switch self {
        case .main: return 1
        //...
        case .taiChi: return 1999
```

and add a new `uniqueChainId` that is exactly the same as `chainID`, except `.taiChi` returns `1`.

And then

`uniqueChainId` is only used when connecting to RPCServers (and specifically, not when signing).

A little impure, but seems like it might work.. What do you think?

**Comment by vladyslav-iosdev** [comment]
> How about we invent a different chainId (say `1999`, it's not listed in https://github.com/ethereum-lists/chains/tree/master/_data/chains) and assign that to TaiChi, so:
> 
> ```
> enum RPCServer... {
>     var chainID: Int {
>         switch self {
>         case .main: return 1
>         //...
>         case .taiChi: return 1999
> ```
> 
> and add a new `uniqueChainId` that is exactly the same as `chainID`, except `.taiChi` returns `1`.
> 
> And then
> 
> `uniqueChainId` is only used when connecting to RPCServers (and specifically, not when signing).
> 
> A little impure, but seems like it might work.. What do you think?

yep, seems like it could be esiest way,, i will check it.

**Comment by hboon** [comment]
This is wrong:

> uniqueChainId is only used when connecting to RPCServers (and specifically, not when signing).

I mean:

uniqueChainId is only used when connecting to RPCServers (and specifically when signing, because we want it to be `1` when signing).

**Comment by hboon** [comment]
@vladyslav-iosdev let me think and maybe clarify more about the feature first. It seems like a better way to do this is to provide a switch/toggle to use TaiChi either in the send screen/actionsheet or in the Settings tab?

**Comment by hboon** [comment]
@zhangzhongnan928 How important is this feature?

It seems like a better way to do this is to provide a switch/toggle to use TaiChi either in the send screen/actionsheet or in the Settings tab [1].

Because if we add it as a new chain, users can only send ETH and not run other transactions like send ERC20 or other smart contract calls with other tokens. Adding a new chain also makes the code and UI inconsistent, since there are now tokens (maybe just ETH) that appear twice in the Wallet tab.

[1] there will be some additional background work to make this work nicely, e.g. we should remember where if the mainnet transaction was posted through TaiChi or not and check the pending status with TaiChi, instead of mainnet.

**Comment by zhangzhongnan928** [comment]
> @zhangzhongnan928 How important is this feature?
> 
> It seems like a better way to do this is to provide a switch/toggle to use TaiChi either in the send screen/actionsheet or in the Settings tab [1].
> 

agree, but too much work, right?

> Because if we add it as a new chain, users can only send ETH and not run other transactions like send ERC20 or other smart contract calls with other tokens. Adding a new chain also makes the code and UI inconsistent, since there are now tokens (maybe just ETH) that appear twice in the Wallet tab.
> 

The key feature is to participate DeFi with private TX. If cannot use for other smart contracts, then useless.
In this case, we can keep it as a feature in the backlog

> [1] there will be some additional background work to make this work nicely, e.g. we should remember if the mainnet transaction was posted through TaiChi or not and check the pending status with TaiChi, instead of mainnet.

Not very important for now.

The goal is:
1. to test the private TX feature
2. to start the relationship with spark pool, in the future, we may use the TaiChi to replace infura or use one as a backup for the other.

in conclusion, we can close it

**Comment by hboon** [comment]
> > @zhangzhongnan928 How important is this feature?
> > It seems like a better way to do this is to provide a switch/toggle to use TaiChi either in the send screen/actionsheet or in the Settings tab [1].
> 
> agree, but too much work, right?

Let's try it out. It's going to work better for the user, and I *think* it's going to be simpler and more maintainable code-wise. We can do this:

* Add a toggle/switch to the gas settings screen (where the user can set the nonce) labelled "Use TaiChi"
* Don't do anything special with pending transactions status. It'll still work because we already check against Infura.

Then we should be able to go ahead with:

> 1. to test the private TX feature
> 2. to start the relationship with spark pool, in the future, we may use the TaiChi to replace infura or use one as a backup for the other.

And if it's good, we can see if we want/how to make it even better.

**Comment by vladyslav-iosdev** [comment]
> * Don't do anything special with pending transactions status. It'll still work because we already check against Infura.

@hboon its not clear what do i need to do here, can u provide more details?

**Comment by hboon** [comment]
Don’t have to do anything for checking/supporting pending status I think. Should just work. Just need to post to the new RPC node
